### PR TITLE
Add `verify_base_db` method to check if all base tables are present.

### PR DESCRIPTION
**Pull request by vedanshujain**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Adding another flag `woocommerce-schema-version`.

One of the notable changes in this PR is that we are creating another flag: `woocommerce-schema-version` to store the version till which we have verified that base schema DB is installed.

Even though we already have similar flags:

1. `woocommerce_version` to store the WooCommerce Version

2. `woocommerce_db_version` to store the version point till where we have run the update function.

I wanted to use `woocommerce_db_version` but it was not suitable because having not run update callbacks is quite different from base schema structure mismatch. Also, I did not want to prevent update callback here because this responsibility of erroring and scheduling themselves again should fall to the callback code itself.

### Changes proposed in this Pull Request:

 Add `verify_base_db` method to check if all base tables are present.

Optionally, it also adds a notice in case all DB tables are not present.

Note that we only check missing tables and don't care about exact table structure because many time tables are modified by merchants to better suit their needs (indexes, collations, etc).

### How to test the changes in this Pull Request:

Part 1

1. Rename a base table, maybe `{prefix}wc_tax_rate_classes` to `wc_tax_rate_classes_2`.
1. Comment out `self::create_tables();` line to make sure table is not created on refresh.
1. Delete option `woocommerce_version` and `woocommerce_db_version` to trigger update flow.
1. You should be able to see notice similar to 
``` 
One or more tables required for WooCommerce to functions are missing, some features may not work as expected. Missing tables: wc_wc_tax_rate_classes. 
```
Don't dismiss the notice just yet.
1. Apply the patch in https://github.com/woocommerce/woocommerce-rest-api/pull/188
1. Rename the table back to `{prefix}wc_tax_rate_classes`.
1. Refresh the page.
1. Click on "Check again" text in the notice.
1. Notice will go away on refresh.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Dev - Add `verify_base_db` method to check if all base tables are present.

### CODE REVIEW

**Comment by Konamiman** [comment]
Before applying the REST API patch:

image

After applying the patch:

image

After renaming back the table, clicking "Check again" and refreshing:

image

So 👍🏻

File: includes/admin/views/html-notice-base-table-missing.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+		<?php
+		$verify_db_tool_available = array_key_exists( 'verify_db_tables', WC_Admin_Status::get_tools() );
+		$missing_tables           = get_option( 'woocommerce-schema-missing-tables' );
+		if ( $verify_db_tool_available ) {
+			echo wp_kses_post(
+				sprintf(
+				/* translators: %1%s: Missing tables (seperated by ",") %2$s: Link to check again */
+					__( 'One or more tables required for WooCommerce to function are missing, some features may not work as expected. Missing tables: %1$s. <a href="%2$s">Check again.</a>', 'woocommerce' ),
+					esc_html( implode( ', ', $missing_tables ) ),
+					wp_nonce_url( admin_url( 'admin.php?page=wc-status&tab=tools&action=verify_db_tables' ), 'debug_action' )
```
Comment (claudiosanches) [inline includes/admin/views/html-notice-base-table-missing.php:30]: Missing `esc_url`.

File: includes/admin/views/html-notice-base-table-missing.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+			<?php esc_html_e( 'Dismiss', 'woocommerce' ); ?>
+	</a>
+
+	<p>
+		<strong><?php esc_html_e( 'Database tables missing', 'woocommerce' ); ?></strong>
+	</p>
+	<p>
+		<?php
+		$verify_db_tool_available = array_key_exists( 'verify_db_tables', WC_Admin_Status::get_tools() );
+		$missing_tables           = get_option( 'woocommerce-schema-missing-tables' );
```
Comment (claudiosanches) [inline includes/admin/views/html-notice-base-table-missing.php:23]: It need to get updated to `woocommerce-schema-missing-tables`, by default on WooCommerce we use `woocommerce_` as a prefix for all options, note that also we remove options with this prefix when running the `unistall.php`.

File: includes/class-wc-install.php [hunk 3]
```diff
@@ -303,6 +304,43 @@ public static function install() {
...
+		if ( 0 < count( $missing_tables ) ) {
+			if ( $modify_notice ) {
+				WC_Admin_Notices::add_notice( 'base_tables_missing' );
+			}
+			update_option( 'woocommerce-schema-missing-tables', $missing_tables );
+		} else {
+			if ( $modify_notice ) {
+				WC_Admin_Notices::remove_notice( 'base_tables_missing' );
+			}
+			update_option( 'woocommerce-schema-version', WC()->version );
```
Comment (claudiosanches) [inline includes/class-wc-install.php:338]: Need to replace `-` for `_` in this option key.

**Review by claudiosanches — CHANGES_REQUESTED** [review]
Looks good, great work and makes sense to save the database schema in a new option.

I placed some notes about tweaks that we could do, but I think we could have a database schema version instead of using WooCommerce's version, now we are registering the WooCommerce version used when the user updated and had the database version updated, but this isn't the database schema version.

File: includes/class-wc-install.php (hunk 3 above)
Comment (vedanshujain) [reply includes/class-wc-install.php]: Changed in https://github.com/woocommerce/woocommerce/pull/26454/commits/4fce106f1446eb5e8fec45bd040b2cacde971880

File: includes/admin/views/html-notice-base-table-missing.php (hunk 1 above)
Comment (vedanshujain) [reply includes/admin/views/html-notice-base-table-missing.php:29]: Changed in https://github.com/woocommerce/woocommerce/pull/26454/commits/4fce106f1446eb5e8fec45bd040b2cacde971880

File: includes/admin/views/html-notice-base-table-missing.php (hunk 2 above)
Comment (vedanshujain) [reply includes/admin/views/html-notice-base-table-missing.php]: Changed in https://github.com/woocommerce/woocommerce/pull/26454/commits/4fce106f1446eb5e8fec45bd040b2cacde971880

File: includes/admin/views/html-notice-base-table-missing.php (hunk 1 above)
Comment (claudiosanches) [reply includes/admin/views/html-notice-base-table-missing.php:29]: WordPress mention that the URL get escaped, but in several places also uses `esc_url()`, on WooCommerce we have the same. But sounds like we don't need to worry about it here.

**Review by claudiosanches — APPROVED** [review]
Great work!

**Pull request merged**